bin/propolis-server/src/lib/migrate/source.rs

```diff
@@ -114,77 +117,106 @@ enum RamOfferDiscipline {
+pub(crate) async fn initiate<T: MigrateConn>(
+    log: &slog::Logger,
+    migration_id: Uuid,
+    mut conn: WebSocketStream<T>,
+    vm_objects: &VmObjects,
+    persistent_state: &PersistentState,
+) -> Result<impl SourceProtocol, MigrateError> {
+    // Create a new log context for the migration
+    let log = log.new(slog::o!(
+        "migration_id" => migration_id.to_string(),
+        "migrate_role" => "source"
+    ));
+    info!(log, "negotiating migration as source");
+
+    let selected = match conn.next().await {
+        Some(Ok(tungstenite::Message::Text(dst_protocols))) => {
+            info!(log, "destination offered protocols: {}", dst_protocols);
+            match protocol::select_protocol_from_offer(&dst_protocols) {
+                Ok(Some(selected)) => {
+                    info!(log, "selected protocol {:?}", selected);
+                    conn.send(tungstenite::Message::Text(
+                        selected.offer_string(),
+                    ))
+                    .await?;
+                    selected
+                }
+                Ok(None) => {
+                    let src_protocols = protocol::make_protocol_offer();
                     error!(
-                        proto.log(),
-                        "failed to restore dirty bits: {e}";
-                        "gpa" => gpa,
+                        log,
+                        "no compatible destination protocols";
+                        "dst_protocols" => &dst_protocols,
+                        "src_protocols" => &src_protocols,
                     );
-                    // No sense continuing to try putting back any remaining
-                    // dirty bits, as we won't be using them any longer.
-                    break;
-                } else {
-                    debug!(proto.log(), "re-dirtied pages at {gpa:#x}",);
+                    return Err(MigrateError::NoMatchingProtocol(
+                        src_protocols,
+                        dst_protocols,
+                    ));
+                }
+                Err(e) => {
+                    error!(log, "failed to parse destination protocol offer";
+                           "dst_protocols" => &dst_protocols,
+                           "error" => %e);
+                    return Err(MigrateError::ProtocolParse(
+                        dst_protocols,
+                        e.to_string(),
+                    ));
                 }
             }
         }
+        x => {
+            conn.send(tungstenite::Message::Close(Some(CloseFrame {
+                code: CloseCode::Protocol,
+                reason: "did not begin with version handshake.".into(),
+            })))
+            .await?;
+            error!(
+                log,
+                "destination side did not begin migration version handshake: \
+                 {:?}",
+                x
+            );
+            return Err(MigrateError::Initiate);
+        }
+    };
```

Comment:
style nit, take it or leave it: there are two `match`es here, one of which handles whether or not we receive a text message on the websocket stream, and another which handles whether the text message can be parsed successfully. currently, the second match is nested inside of the first one. IMHO, this code would be a lot easier to read if the two matches were in sequence, rather than nested inside of each other, like this: 
```rust

    let dst_protocols = match conn.next().await {
        Some(Ok(tungstenite::Message::Text(dst_protocols))) => dst_protocols,
        x => {
            conn.send(tungstenite::Message::Close(Some(CloseFrame {
                code: CloseCode::Protocol,
                reason: "did not begin with version handshake.".into(),
            })))
            .await?;
            error!(
                log,
                "destination side did not begin migration version handshake: \
                 {:?}",
                x
            );
            return Err(MigrateError::Initiate);
        },
    };

    info!(log, "destination offered protocols: {}", dst_protocols);
    let selected = match protocol::select_protocol_from_offer(&dst_protocols) {
        Ok(Some(selected)) => {
            info!(log, "selected protocol {:?}", selected);
            conn.send(tungstenite::Message::Text(
                selected.offer_string(),
            ))
            .await?;
            selected
        }
        Ok(None) => {
            let src_protocols = protocol::make_protocol_offer();
            error!(
                log,
                "no compatible destination protocols";
                "dst_protocols" => &dst_protocols,
                "src_protocols" => &src_protocols,
            );
            return Err(MigrateError::NoMatchingProtocol(
                src_protocols,
                dst_protocols,
            ));
        }
        Err(e) => {
            error!(log, "failed to parse destination protocol offer";
                    "dst_protocols" => &dst_protocols,
                    "error" => %e);
            return Err(MigrateError::ProtocolParse(
                dst_protocols,
                e.to_string(),
            ));
        }
    };
```
(I couldn't leave this as a "suggestion" because it contains deleted lines apparently :/)

This way, we are saying "first, try to receive a text message, or bail out if we get something weird. then, try to parse that message". I think that's a bit easier to follow for the reader.